File: FlattenNestedListIterator_BruteForce.py
# TC: O(N) where N is the number of lists in the nested list. Iterating over them recursively will take O(N) time. 
# SC: O(D) where D is the maximum depth of the recursion, which will also be the height of the stack. 

# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger:
#    def isInteger(self) -> bool:
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        """
#
#    def getInteger(self) -> int:
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        """
#
#    def getList(self) -> [NestedInteger]:
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        """

class NestedIterator:
    def __init__(self, nestedList: [NestedInteger]):
        def flattenList(int_list): 
            for i in int_list: 
                if i.isInteger(): 
                    self.int_list.append(i)
                else: 
                    flattenList(i.getList())
            
        self.int_list = []
        self.curr = -1
        flattenList(nestedList)
    
    def next(self) -> int:
        self.curr += 1
        return self.int_list[self.curr]
          
    def hasNext(self) -> bool:
        if self.curr + 1 < len(self.int_list):
            return True
        return False
              

# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: FlattenNestedListIteratorBruteForce.java
// Brute force approach: We use dfs to get all integers in our result in order. If in our dfs, we encounter an integer, we simply add it to the result. If it is not an integer and a list instead, we run dfs on that list. The problem with this approach is that it is creating the result all at once. An iterator should just care about the next element and should go one by one. If there are any chnages to the list, this iterator will not be able to handle that which defeats the purpose of having an iterator

// Space complexity: O(N + D). The N is for the result array list and D is for the recursion stack where D is the maximum nesting depth


public class NestedIterator implements Iterator<Integer> {
    List<Integer> result;
    int index = 0;
    public NestedIterator(List<NestedInteger> nestedList) { // Time: O(N + L) due to time taken to build the result array list
        result = new ArrayList<>();
        dfs(nestedList);
    }

    @Override
    public Integer next() { // Time: O(1)
        return result.get(index++);
    }

    @Override
    public boolean hasNext() { // Time: O(1)
        return index != result.size();
    }
    
    private void dfs(List<NestedInteger> nestedList) {
        for (NestedInteger ni : nestedList) {
            if (ni.isInteger()) {
                result.add(ni.getInteger());
            }
            else {
                dfs(ni.getList());
            }
        }
    }
}

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */